site/_posts/2024-11-09-fish-4beta-highlights.md

```diff
@@ -0,0 +1,247 @@
+We hope to be both contributing to and benefitting from future standardization.
+
+Since fish is not usually preinstalled, it has a lot of incentive to play nice with other systems.
+In the past fish has done some user-agent sniffing by enabling or disabling code based on `$TERM`.
+We've removed a lot of these workarounds already,
+to reduce surprise and encourages fixing bugs at the source,
+which helps our peers that don't have our workarounds.
+
+Among others, fish implements
+
+- Operator System Command (OSC) 0 terminal title
+- OSC 7 working directory reporting
+- OSC 52 copy to clipboard
+- OSC 133 prompt marking
+- (partially) setting/restoring the terminal cursor
+- (partially) focus reporting
+- Parts of the [kitty keyboard protocol](https://sw.kovidgoyal.net/kitty/keyboard-protocol/)
+
+## Under the Hood
+
+The remaining sections look at the recent Rust port and some related topics regarding internal development.
+
+### Why port from C++ to Rust
+
+Fish was implemented in C++11 and has recently been ported to [Rust](https://www.rust-lang.org/).
+
+I love C++.
+Reading and writing C++ is like standing on the shoulders of giants.
+
+Unfortunately, it seems to become increasingly hard to find C++ programmers motivated to work on shiny tools,
+especially within fish's target audience.
+fish is a project with a large scope, attempting to solve lots of UI problems by itself.
+It is written by hundreds volunteers who, over the years have come and gone,
+leaving behind a heterogenous code base.
+Our low requirements on the quality of contributions enable growth but also add tech debt.
+Mainstream programming styles have changed a lot since fish was initially written almost 20 years ago.
+It doesn't help that we were stuck on the fairly old C++11 and afraid to upgrade for some reason.
+
+It is definitely possible to make [basically finished software out of beautiful C++20](https://kakoune.org/)
+but given fish's aspirations, historical baggage and present resources, that's unrealistic.
+
+To enable the project to continue to grow at a solid pace, we
+[ported it to Rust](https://github.com/fish-shell/fish-shell/pull/9512).
+Rust has definitely become the language of choice for comparable projects.
+Some highlights are its modern defaults, impressive standard library,
+a coherent ecosystem and a great story on multi-threading[^multi-threading].
+
+### Porting Process
+
+This was an incremental port, using [cxx](https://cxx.rs/) and [autocxx](https://google.github.io/autocxx/) for interop.
+These tools have known limitations, namely that function calls crossing the
+foreign-function-interface (FFI) boundary can pass only certain types as arguments.
+
+To port a piece of C++ code, the steps are:
+1. Find all dependencies of the code, ideally port them first.
+   Doing so avoids the need to call C++ functions from Rust,
+   which can be tricky due to
+   which might need work around limitations of the FFI.
```

Comment:
I'm unsure what you wanted to say, but `need work` is invalid

Is it one of these ?

```suggestion
   which might need workaround limitations of the FFI.
```

```suggestion
   which might need to work around limitations of the FFI.
```

or something else ?

Maybe I'm simply wrong :sweat_smile: